util/repr-bytes/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+#[derive(Debug)]
+pub struct LengthMismatch {
+    expected: usize,
+    found: usize,
+}
+
+impl Display for LengthMismatch {
+    fn fmt(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(
+            formatter,
+            "Expected exactly {} bytes, got {}",
+            self.expected, self.found
+        )
+    }
+}
+
+////
+// Suggested Implementations:
+// These macros provide instances of what can be thought of as "blanket implementations"
+// on a per-type basis.
+// These are macros because rust's coherence rules mean that blanket implementations
+// won't work out well for this use-case.
+//
+// There are two types of suggested impls:
+// - Impls of other traits in terms of ReprBytes
+// - Impls of ReprBytes in terms of other traits
+////
+
+/// Derive ReprBytes from AsRef<[u8]>, TryFrom<&[u8]>, and Size as a typenum.
+/// This is expected to be the right implementation for almost all cryptographic
+/// primitives, e.g. X25519, CompressedRistretto, etc.
+/// It can't work for e.g. RistrettoPoint, which doesn't have AsRef<[u8]>.
+///
+/// Arguments:
+///   - $mytype is the type you want to impl ReprBytes
+///   - $mysize is a typenum, representing the size of the canonical representation
+///
+/// Requirements:
+///   - <AsRef<[u8]> for $mytype>::as_ref().len() always equals $mysize::USIZE
+///   - <TryFrom<&[u8]> for $mytype>::Error implements core::fmt::Display
+///   - <TryFrom<&'a[u8]> for $mytype>::Error is the same for all values of 'a,
+///     OR they are all convertible to the value when 'a = 'static, via core::convert::From.
+#[macro_export]
+macro_rules! derive_repr_bytes_from_as_ref_and_try_from {
+    ($mytype:ty, $mysize:ty) => {
+        impl $crate::ReprBytes for $mytype {
+            type Size = $mysize;
+            type Error = <$mytype as TryFrom<&'static [u8]>>::Error;
+
+            fn from_bytes(src: &$crate::GenericArray<u8, Self::Size>) -> Result<Self, Self::Error> {
+                Ok(<Self as TryFrom<&[u8]>>::try_from(src.as_slice())?)
+            }
+
+            fn to_bytes(&self) -> $crate::GenericArray<u8, Self::Size> {
+                use $crate::{typenum::Unsigned, GenericArray};
+                let slice = <Self as AsRef<[u8]>>::as_ref(self);
+                debug_assert!(slice.len() == <Self::Size as Unsigned>::USIZE);
+                <GenericArray<u8, Self::Size>>::from_slice(slice).clone()
+            }
+
+            // Optimization: Use AsRef<[u8]> here, which also makes Into<Vec<u8>> better
+            fn map_bytes<T, F>(&self, f: F) -> T
+            where
+                F: FnOnce(&[u8]) -> T,
+            {
+                f(<Self as AsRef<[u8]>>::as_ref(self))
+            }
+        }
+    };
+}
+
+/// Derive Into<Vec<u8>> from a ReprBytes implementation
+#[macro_export]
+macro_rules! derive_into_vec_from_repr_bytes {
+    ($mytype:ty) => {
+        impl Into<::alloc::vec::Vec<u8>> for $mytype {
```

Comment:
Same story here, in case the crate downstream of us isn't doing it's own `extern crate alloc;` directly.